framework/src/preference/preference_get.c

```diff
@@ -0,0 +1,193 @@
+/****************************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <stdio.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <debug.h>
+#include <preference/preference.h>
+
+#include "preference_internal.h"
+
+static int preference_fs_read(char *key, preference_data_t *data)
+{
+	int ret;
+	int type;
+	char *path;
+
+	if (key == NULL || data == NULL) {
+		prefdbg("Invalid parameter\n");
+		return PREFERENCE_INVALID_PARAMETER;
+	}
+
+	ret = preference_get_keypath(key, &path);
+	if (ret < 0) {
+		prefdbg("Failed to get preference path\n");
+		return PREFERENCE_PATH_NOT_FOUND;
+	}
+
+	int fd = open(path, O_RDONLY, 0666);
+	free(path);
+
+	if (fd < 0) {
+		if (errno == ENOENT) {
+			return PREFERENCE_INVALID_PARAMETER;
+		}
+		return PREFERENCE_IO_ERROR;
+	}
+
+	/* Read the type of data value and Compare to the type of request */
+	ret = read(fd, (void *)&type, sizeof(int));
+	if (ret < 0) {
+		ret = PREFERENCE_IO_ERROR;
+		goto errout;
+	} else if (type < 0 || data->type < 0 || data->type != type) {
```

Comment:
Type ranges from 0 to 4. Only checking negative value may be not enough.

Reply:
That value is checked in switch context below.

Reply:
If so, it's kind of duplicated checking. That is, only checking 'data->type != type' may be enough. Just my opinion.